### ISSUE TITLE
Additional display options for Stereoscopic 3D

### ISSUE THREAD

**Issue by NeatNit** [citra-emu/citra#4255]
Some of this is discussed in the original pull request #3632, but I thought it'd be better to open a new issue and list the settings I think are missing.

### Option to switch left and right images
This will make it possible to see 3D using the [cross-eyed method](http://www.vision3d.com/3views.html). Currently, the image intended for the left eye is shown on the left side of the screen, and the right eye one on the right. The cross-eyed method requires them to be reversed.

### Option to disable horizontal squashing
The current output is currently squashed to half its width, which is how 3D monitors and TVs often expect to get a side-by-side 3D image. Again, this is bad for the cross-eyed method and even the parallel method, as both methods depend on having two natural-looking images side by side.

### Different 3D output options
The following methods (see also: http://www.stereo3d.com/formats.htm) are used by various different 3D display technologies:
1. Side by side (only method currently implemented in Citra)
2. Top and bottom - for many passive 3D displays, this would result in no data loss, while side by side has 50% of the pixels discarded by the monitor.
3. Line interlacing - odd lines are left image & even lines are right image, or vice versa. Used by many passive 3D displays.
4. Column interlacing - I'm actually not sure if this is actually used by any technology (other than the 3DS itself!), but if you implement line interlacing I would think that column interlacing would be trivial to add. Better safe than sorry!
5. Frame stacking / page flipping - every monitor refresh switches to the other eye. This is used by active shutter 3D glasses, such as Nvidia 3D. Of course, switching back and forth between the two images must be v-synced and independent of the game's framerate.

I ordered them in the order of what I think would be easiest to implement to hardest. As it always is, the frame stacking method (hardest) is probably the most important / would have the most users.

**Comment by NeatNit** [comment]
Also, whatever this guy is asking (I couldn't decipher it): #4115

**Comment by NeatNit** [comment]
I added Anaglyph 3D to the list, and linked to some relevant Wikipedia pages.

**Comment by MokuDuk** [comment]
Anaglyph 3D is very fun to play. [especially by controlling display color temperature with f.lux or redshift]

Even on existing much slower citra builds there was a Stereoscopic Slider, which enhanced native 3ds graphic quality like 1000 times. And there is even artists\designers who like to watch anaglyph effect with there own bare pair of eyes with no glasses at all.

Dolphin emulator has the best anaglyph support nowadays. In pcsx2 there is an old and very beautiful graphical plugin for anaglyph support too [its abandoned now and requires updates for new fixes coming from main development plugin]

Also if Citra will ever be working in DirectX 9, 10 or 11 backend - stereoscopy could be enabled by Tridef or iz3D. Both software giving different results in performance & visuals and working like a charm in ppsspp and in new native windows games. [I also remember some long way to make it work for DeSmuME, pj64 or mupen64]

And finally it might be working with compiz anaglyph support even now [full 3D stereoscopy? Or flat image photo\video stereoscopy?]. [I will install & configure linux only after great amount of time]

**Comment by DarthJahus** [comment]
I'd like to add my voice to this feature request.

* Anaglyph 3D (red/cyan channels) would allow everyone to experience 3D without any additional hardware, just cheap or free stereoscopic glasses. It's made just by superimposing the 2 already generated images, one without the red channel, and the other without blue and green channels.

* Removing the stretch from images would allow us to appreciate the full experience, just by crossing the eyes.

* Allowing to switch left and right images would allow Android users to experience 3D with a cardboard.

**Comment by JellyDean446** [comment]
They is a better branch now / and on page 458 new build ^

https://github.com/citra-emu/citra/pull/4578

drwhojan

**Comment by zhaowenlan1779** [comment]
Ugh, please do not post unofficial builds here. thanks a lot

**Comment by JellyDean446** [comment]
Ok ^, But why keep marking my post as spam ?.., clearly most ant.

**Comment by DarthJahus** [comment]
@JellyDean446, because you are not allowed to post unofficial builds here. Same reason why @MokuDuk's comment is hidden.

**Comment by JellyDean446** [comment]
> @JellyDean446, because you are not allowed to post unofficial builds here. Same reason why @MokuDuk's comment is hidden.

Ok thanks..

**Comment by FearlessTobi** [comment]
Keep the convo on topic and within our rules or I'll lock this.

**Comment by masterotaku** [comment]
Today I have made a Citra SBS to 3D Vision conversion. Instructions and download link here: https://helixmod.blogspot.com/2019/08/citra-nintendo-3ds-emulator-sbs-to-3d.html
Direct link to the file (but people should really read the instructions): https://masterotaku.s3.amazonaws.com/Citra/Citra_3D_Vision_fix.7z

In short, it resizes the image to the original aspect ratio (which moves the right half outside of the screen) and then for the right eye it places the correct image where it should be. Because of that resizing, I recommend using 4x DSR to avoid pixelation. And I made an extra feature: being able to change the depth of the game screens.

From the modes list in the OP, number 6 is now covered, in a workaround way. Only for Nvidia GPUs using drivers that still contain 3D Vision.

Thanks to the OpenGL wrapper I used (source: http://3dsurroundgaming.com/OGL3DVision.html), I didn't need a DX11 renderer in the end. Older versions of the wrapper caused crashes a year ago, and that's when I was frustrated. I just needed to wait for the right time.

**Comment by Kzitold** [comment]
_bump_

Need to un-squish side-by-side stereoscopy.

**Comment by lividhen** [comment]
> 
> 
> _bump_
> 
> Need to un-squish side-by-side stereoscopy.

yeah... the squashed screens make it difficult to read text.

**Comment by Kotters** [comment]
I'd also like to voice a request for unsquished, Full-SBS output in 3D. With the advent of VR headsets, this is one of the best and most accessible ways to play 3D content. As VR headsets provide a full display for each eye, there's no need to horizontally squish the image.

**Comment by bafterfor** [comment]
I also am looking for a non-stretched option.

**Comment by DarthJahus** [comment]
It would be good to have an option to run the game with OpenVR.

**Comment by jvook** [comment]
Can we have an option to select the anaglyph colors.  I have green magenta glasses which are better for color vision.

**Comment by jakedowns** [comment]
> Can we have an option to select the anaglyph colors? I have green magenta glasses which are better for color vision.

here's some shaders for some additional color options
https://www.reddit.com/r/emulation/comments/t0ejg5/more_anaglyph_3d_shaders_for_dolphin_and_citra/

**Comment by jakedowns** [comment]
I'm gonna look into an un-squishing option :) I'll toss in a swap L/R option while i'm at it

fwiw, in the meantime, if you want to test my reshade effect, it has an option for swapping l/r already:
https://github.com/jakedowns/reshade-shaders/blob/dual-buffer-plus-rgb/Citra%20AddOn/Citra.fx

you just check "preview depth" "swap l/r" and lower "preview alpha" to 0

**Comment by jakedowns** [comment]
still need to get the other layout modes working. but im making progress on unsquishing aka SBS - Full

https://user-images.githubusercontent.com/1683122/195957599-36a584c4-4f1b-4712-8e89-9daa7e74d8f4.MOV

**Comment by logicallysound** [comment]
Has there been further progress here? Full SBS would be a big boon for Nreal glasses users. Nreal recently released a beta firmware that allows for stereoscopic 3D but it actually has a full 1920x1080@60 display for each eye so the current SBS output in Citra nightly appears squished.

**Comment by jakedowns** [comment]
ive been meaning to pick this up again recently. been meaning to add a "separation" slider for nreal air specifically. ill see if i can make any progress on it this week

**Comment by nathanbain** [comment]
Looking forward to this. 3d gaming on the Nreal is amazing, and fixing the "squish" problem would make it even better!

**Comment by Mad-Matty** [comment]
Would it be possible to add a Native 800x240 for use with the side by side 3d mode giving the full 400x per eye.  This would avoid the vertical stretching and no need for horizontal stretching of an already cut down 200x per eye.  Having a swap L/R function would be handy for the crosseye effect as well.

**Comment by jakedowns** [comment]
ive been meaning to wrap up this asap. i have it on a feature branch on my fork, i just need to get it up to date with the latest mainline code, test, and make the pr.

**Comment by UncleUncleWang** [comment]
> ive been meaning to wrap up this asap. i have it on a feature branch on my fork, i just need to get it up to date with the latest mainline code, test, and make the pr.

hi jake，Is there any new progress on the fix of sbs? Is there a version that can be experienced, thank you very much. I recently got an nreal air and want to test the effect

**Comment by jvook** [comment]
> > Can we have an option to select the anaglyph colors? I have green magenta glasses which are better for color vision.
> 
> here's some shaders for some additional color options https://www.reddit.com/r/emulation/comments/t0ejg5/more_anaglyph_3d_shaders_for_dolphin_and_citra/

Thank you very much for replying and sharing this.  But it is actually kind of funny, because those are my files and I posted that because I never found anything :)

**Comment by lividhen** [comment]
> > > Can we have an option to select the anaglyph colors? I have green magenta glasses which are better for color vision.
> > 
> > here's some shaders for some additional color options https://www.reddit.com/r/emulation/comments/t0ejg5/more_anaglyph_3d_shaders_for_dolphin_and_citra/
> 
> Thank you very much for replying and sharing this.  But it is actually kind of funny, because those are my files and I posted that because I never found anything :)
> 
> I'm actually going to be deleting all my reddit content due to some major anti-consumer stuff they are doing over there (basically I just don't want them earning money off my stuff anymore).  So I'll just post here that the link to those files is https://www.dropbox.com/s/ohvhqflezog325v/ffmpeg-anaglyph-shaders.zip?dl=0 , and it includes several shader variations for each of the 3-D glasses colors.  I think it has a readme file with instructions.

Will you be moving to another service like kbin / lemmy? I have moved some of my posts to the respective communities over there so people can still find them in the future if they need the info.

**Comment by corlin-fardal** [comment]
Has anyone gotten the linked anaglyph shaders to actually work in Citra? They work just fine in Dolphin, but I know Citra tends to be picky about shaders, and it's also not going into anaglyph mode at all after booting a game if I don't change the screen mode to reset the screen first. Is there some setting or something I'm missing to make the shaders work, or do they just not?

**Comment by MokuDuk** [comment]
[> Has anyone gotten the linked anaglyph shaders to actually work in Citra?

extract archive, place files in \user\shaders\anaglyph. Tried yesterday on unofficial 2018 build, everything works amazingly except that "fullcolor" at least for Red/Green is oversaturated in comparison with Reshade and other Tridef, iz3d methods. Maybe it's easy to refine just with Notepad, haven't tried yet, because don't play everyday, whole world/planet nowadays in danger, you know]

no one stated an aim, and maybe it's important - dolphin, pcsx2, ppsspp are "forcing" stereoscopic output everywhere, even if it wasn't on console. So in case of 3ds it would be cool to see both 3ds screens in stereo, especially when other emulators do this. And even when they don't - Reshade so-called "fake" Depth3D methods are working everywhere, including even some software [they work with citra too]. Also, correct me if I'm wrong, some video games when turning on stereo output - are disabling posteffects [Ridge Racer?]. So if stereo will be bypassed and forced then minor visual effects will retain too

and just to mention - anaglyph[Nvidia vision "Discover" glasses] are different and one made of plastic might be clearer then another one made of paper. And what is more important, they don't cut half of refresh rate, half of resolution, and thus having less input lag unlike other stereoscopic methods. And even cheapest ones can be adjusted for less headache or eyes pain with f.lux, redshift and Reshade's Color Gamut

["normal" - is Reshade's Depth3D]
![normal](https://github.com/citra-emu/citra/assets/20445390/be9f33dd-6d74-4c6b-bca9-b07162609cf9)
![oversaturated](https://github.com/citra-emu/citra/assets/20445390/526173d2-e2f4-402a-b426-98038264854b)

**Comment by corlin-fardal** [comment]
Yeah, I have it in the shaders/anaglyph folder, that's how I'm able to select the shaders with anaglyph 3D enabled. It's good to know that there definitely is one version of Citra that lets the shaders work, but I'd much rather use a recent, official version if at all possible. Can someone test the latest Citra build to confirm it's not just somehow my computer acting strange?

**Comment by MokuDuk** [comment]
> Can someone test the latest Citra build to confirm it

in latest build for Android everything works, it doesn't have an option to choose anaglyph shader other then dubois through UI, but it's possible to specify it in configuration file with any notepad

try:
1) Launch a game
2) Go to settings and no matter what a state enable again anaglyph, your selected shader from an archive, 100% depth [after a game already launched]
3) If it didn't help go to controls, remember or assign keys to change layout, fullscreen; make sure that the game is already launched [it's important] and then test these 2 buttons too

last version for windows 7 of this year of May works with archive from above too

just to mention in general:
1. Android has 3 minor bugs: [or my phone Android 13 Vulkan default system drivers]
1.1. Anaglyph output doesn't turn on until you launch a game and go to settings bang it over any of it's off/on state [ppsspp has the same, so not sure, is it emulator or just default system drivers]
1.2. In single screen mode "swap screens" button changes layout instead of swapping. Only on first press, second and further are working as intended
1.3. pause button works only once. Crashes if used more. Going anywhere out of emulator loses being paused and if you try to pause it over loosen control- crashes again

2. Android missing 2 minor features of desktop:
2.1. Selecting shader other then "dubois" in UI [from archive above; if edited/specified with notepad - everything is amazing]
2.2. Rotating upright [or in other words arrows/touch controls - portrait; graphical output - landscape]

3. All versions. With editing of conf file it's possible to go beyond 100% [225%] of depth, which is personally find cooler, and maybe such ability is better to have in UI too

all these notes are minor, Citra is amazing and with Vulkan it unbelievably runs on Android not worse then with GL on Windows, even in real/true and specifically designed stereoscopy, which is known to half the performance

**Comment by corlin-fardal** [comment]
Switching it from Vulkan to OpenGL on PC is what made it work, thank you for that reminder! That really should have occurred to me sooner, but it slipped my mind as a thing to check for for whatever reason

**Comment by jakedowns** [comment]
guessing we should close this in favor of https://github.com/citra-emu/citra/issues/6826

**Comment by NeatNit** [comment]
> guessing we should close this in favor of #6826

I won't mind if a maintainer closes it, but I'll let them decide.
